wgengine/magicsock: crash

My `tailscaled` crashed after the recent magicsock refactoring work.

I can't figure out exactly which commit it was, but something recent.  /cc @danderson 

```
magicsock: created endpoint key=[8xcuW]: disco=d:bf4e3f302e6ea487; derp=2(sfo) aip=100.86.113.101 aip=fd7a:115c:a1e0:ab12:4843:cd96:6256:7165 ep=24.4.62.176:56232 ep=135.180.175.246:39811 ep=[2602:248:7b4a:ff60:6080:26b:f34f:6a30]:50828 ep=10.1.10.240:56232 ep=[2602:248:7b4a:ff60::1111]:56232 ep=[2602:248:7b4a:ff60:18b6:4f2c:c627:a466]:56232 ep=[2602:248:7b4a:ff60:6080:26b:f34f:6a30]:56232
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x8600c2]

goroutine 141824 [running]:
tailscale.com/wgengine/magicsock.(*peerMap).deleteDiscoEndpoint(0xc0001315e8, 0xc000312480)
        /home/bradfitz/src/tailscale.com/wgengine/magicsock/magicsock.go:195 +0xc2
tailscale.com/wgengine/magicsock.(*Conn).SetNetworkMap.func2(0xc000312480)
        /home/bradfitz/src/tailscale.com/wgengine/magicsock/magicsock.go:2147 +0x71
tailscale.com/wgengine/magicsock.(*peerMap).forEachDiscoEndpoint(0x9e6f80, 0xc000603e10)
        /home/bradfitz/src/tailscale.com/wgengine/magicsock/magicsock.go:149 +0x98
tailscale.com/wgengine/magicsock.(*Conn).SetNetworkMap(0xc000131400, 0xc0007002c0)
        /home/bradfitz/src/tailscale.com/wgengine/magicsock/magicsock.go:2145 +0x10fb
tailscale.com/wgengine.(*userspaceEngine).SetNetworkMap(0xc000130500, 0xc0007002c0)
        /home/bradfitz/src/tailscale.com/wgengine/userspace.go:1210 +0x45
tailscale.com/wgengine.(*watchdogEngine).SetNetworkMap.func1()
        /home/bradfitz/src/tailscale.com/wgengine/watchdog.go:108 +0x2c
tailscale.com/wgengine.(*watchdogEngine).watchdog.func1()
        /home/bradfitz/src/tailscale.com/wgengine/watchdog.go:72 +0x1b
tailscale.com/wgengine.(*watchdogEngine).watchdogErr.func1()
        /home/bradfitz/src/tailscale.com/wgengine/watchdog.go:54 +0x29
created by tailscale.com/wgengine.(*watchdogEngine).watchdogErr
        /home/bradfitz/src/tailscale.com/wgengine/watchdog.go:53 +0xb8
```

--- a/wgengine/magicsock/magicsock.go
+++ b/wgengine/magicsock/magicsock.go
@@ -189,9 +189,14 @@ func (m *peerMap) deleteDiscoEndpoint(ep *endpoint) {
 		return
 	}
 	ep.stopAndReset()
-	pi := m.byDiscoKey[ep.discoKey]
+	pi := m.byNodeKey[ep.publicKey]
 	delete(m.byDiscoKey, ep.discoKey)
 	delete(m.byNodeKey, ep.publicKey)
+	if pi == nil {
+		// Kneejerk paranoia from earlier issue 2801.
+		// Unexpected. But no logger plumbed here to log so.
+		return
+	}
 	for ip := range pi.ipPorts {
 		delete(m.byIPPort, ip)
 	}

